File: Python/0046. Permutations.py
# Given a collection of distinct integers, return all possible permutations.
# Example:
# Input: [1,2,3]
# Output:
# [
#   [1,2,3],
#   [1,3,2],
#   [2,1,3],
#   [2,3,1],
#   [3,1,2],
#   [3,2,1]
# ]
# https://leetcode.com/problems/permutations/description/
# 
# Author: Adam Yu Wen
# Date: May 21, 2018

class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        visit = [0] * len(nums)
        out, res = [], []
        self.permuteDFS(nums, 0, visit, out, res)
        return res
        
    def permuteDFS(self, nums, level, visit, out, res):
        """
        :type nums: List[int]
        :type level: int
        :type visit: List[int]
        :type out: List[int]
        :type res: List[List[int]]
        :rtype: void
        """
        if len(nums) == level:
            # it is import to have [:], which means appending the list
            res.append(out[:])
            return
        else:
            for i in range(len(nums)):
                if visit[i] == 0:
                    visit[i] = 1
                    out.append(nums[i])
                    self.permuteDFS(nums, level + 1, visit, out, res)
                    del(out[-1])
                    visit[i] = 0

File: C++/0046. Permutations.cpp
/* Given a collection of distinct integers, return all possible permutations.
 * Example:
 * Input: [1,2,3]
 * Output:
 * [
 *   [1,2,3],
 *   [1,3,2],
 *   [2,1,3],
 *   [2,3,1],
 *   [3,1,2],
 *   [3,2,1]
 * ]
 * https://leetcode.com/problems/permutations/description/
 * Reference: http://www.cnblogs.com/grandyang/p/4358848.html
 * 
 * Author: Adam Yu Wen
 * Date: May 21, 2018
 */

class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> out;
        vector<int> visited(nums.size(), 0);
        permuteDFS(nums, 0, visited, out, res);
        return res;
    }
    
    void permuteDFS(vector<int> &nums, int level, vector<int> &visited,
		    vector<int> &out, vector<vector<int>> &res) {
        if (level == nums.size()) {
            res.push_back(out);
        } else {
            for (int i = 0; i < nums.size(); i++) {
                if (visited[i] == 0) {
                    visited[i] = 1;
                    out.push_back(nums[i]);
                    permuteDFS(nums, level + 1, visited, out, res);
                    out.pop_back();
                    visited[i] = 0;
                }
            }
        }
    }
};
